lib/krane/global_deploy_task.rb

```diff
@@ -0,0 +1,173 @@
+# frozen_string_literal: true
+require 'tempfile'
+
+require 'krane/common'
+require 'krane/concurrency'
+require 'krane/resource_cache'
+require 'krane/kubectl'
+require 'krane/kubeclient_builder'
+require 'krane/cluster_resource_discovery'
+require 'krane/template_sets'
+require 'krane/resource_deployer'
+require 'krane/kubernetes_resource'
+%w(
+  custom_resource
+  custom_resource_definition
+).each do |subresource|
+  require "krane/kubernetes_resource/#{subresource}"
+end
+
+require 'krane/global_deploy_task_config_validator'
+require 'krane/concerns/template_reporting'
+
+module Krane
+  # Ship global resources to a context
+  class GlobalDeployTask
+    extend Krane::StatsD::MeasureMethods
+    include Krane::TemplateReporting
+    delegate :context, :logger, to: :@task_config
+
+    # Initializes the deploy task
+    #
+    # @param context [String] Kubernetes context
+    # @param global_timeout [Integer] Timeout in seconds
+    # @param selector [Hash] Selector(s) parsed by Krane::LabelSelector
+    # @param template_paths [Array<String>] An array of template paths
+    def initialize(context:, global_timeout: nil, selector: nil, filenames: [])
+      template_paths = filenames.map { |path| File.expand_path(path) }
+
+      @task_config = ::Krane::TaskConfig.new(context, nil)
+      @template_sets = ::Krane::TemplateSets.from_dirs_and_files(paths: template_paths,
+        logger: @task_config.logger)
+      @global_timeout = global_timeout
+      @selector = selector
+    end
+
+    # Runs the task, returning a boolean representing success or failure
+    #
+    # @return [Boolean]
+    def run(*args)
+      run!(*args)
+      true
+    rescue Krane::FatalDeploymentError
+      false
+    end
+
+    # Runs the task, raising exceptions in case of issues
+    #
+    # @param verify_result [Boolean] Wait for completion and verify success
+    # @param prune [Boolean] Enable deletion of resources that match the provided
+    #  selector and do not appear in the template dir
+    #
+    # @return [nil]
+    def run!(verify_result: true, prune: true)
+      start = Time.now.utc
+      logger.reset
+
+      logger.phase_heading("Initializing deploy")
+      validator = validate_configuration
+      resources = discover_resources
+      validator.validate_resources(resources, @selector)
```

Comment:
Not 100% sold that `validate_resources` belongs to the (Global)DeployTaskConfigValidator. But I do like the idea of moving validation code out of this file.